File: selph/multitree.py
"""Multi-tree synthesis for SELPH.

Allows the synthesizer to search across multiple arbitrary namespace trees
simultaneously. Each tree contributes callable components to the search
space, with paths preserved for traceability.

A program synthesized from multiple trees might look like:
  (lambda (x) (ns-get math-lib "double" ((ns-get str-lib "clean") x)))

But for efficiency, we resolve namespace lookups at synthesis time
and generate direct calls:
  (lambda (x) (math.double (str.clean x)))

Usage:
  result = multi_synthesize(
      spec,
      trees=[ops_tree, data_tree, cache_tree],
      tree_names=["ops", "data", "cache"],
  )
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any
from .ast import Node, Symbol, Number, String, List, Spec, GoalExamples
from .eval import (
    Namespace, Closure, BuiltinFn, Env, standard_env,
    eval_node, apply_fn, EvalError,
)
from .synthesize import (
    synthesize, SynthesisResult, Component, _default_components,
    _infer_literal_type, make_validation_fn,
)
from .types import Type, TFn, TNum, TStr, TBool, TList, TVar


# ── Extract components from a Namespace ──────────────────────────────

@dataclass
class TreeComponent(Component):
    """A component extracted from a namespace tree.

    Carries the source tree name and path for traceability.
    """
    tree_name: str = ""
    path: list[str] = field(default_factory=list)


def extract_components(ns: Namespace, tree_name: str = "",
                       prefix: str = "", env: Env | None = None,
                       max_depth: int = 5) -> list[TreeComponent]:
    """Extract callable components from a namespace tree.

    Recursively walks the namespace. For each callable leaf (Closure,
    BuiltinFn), creates a Component with the appropriate type signature.
    For non-callable leaves (numbers, strings), creates constant components.

    The component's node is a Symbol that will be registered in the
    eval environment so it can be called directly.
    """
    if env is None:
        env = standard_env()

    components = []
    _extract_recursive(ns, tree_name, prefix, env, components, 0, max_depth)
    return components


def _extract_recursive(ns: Namespace, tree_name: str, prefix: str,
                       env: Env, components: list, depth: int, max_depth: int):
    """Recursively extract components from a namespace."""
    if depth > max_depth:
        return

    for key, val in ns.items():
        path_str = f"{prefix}.{key}" if prefix else key
        full_name = f"{tree_name}.{path_str}" if tree_name else path_str
        path = path_str.split(".")

        if isinstance(val, Namespace):
            _extract_recursive(val, tree_name, path_str, env, components, depth + 1, max_depth)

        elif isinstance(val, (Closure, BuiltinFn)):
            # Infer type from the callable
            comp_type = _infer_callable_type(val, env)
            if comp_type is not None:
                # Register in env so the synthesizer can call it by name
                env.define(full_name, val)
                components.append(TreeComponent(
                    name=full_name,
                    node=Symbol(full_name),
                    type=comp_type,
                    arity=len(comp_type.params) if isinstance(comp_type, TFn) else 0,
                    tree_name=tree_name,
                    path=path,
                ))

        elif isinstance(val, (int, float)):
            components.append(TreeComponent(
                name=full_name,
                node=Number(float(val)),
                type=TNum(),
                arity=0,
                tree_name=tree_name,
                path=path,
            ))

        elif isinstance(val, str):
            components.append(TreeComponent(
                name=full_name,
                node=String(val),
                type=TStr(),
                arity=0,
                tree_name=tree_name,
                path=path,
            ))

        elif isinstance(val, bool):
            components.append(TreeComponent(
                name=full_name,
                node=Symbol("true") if val else Symbol("false"),
                type=TBool(),
                arity=0,
                tree_name=tree_name,
                path=path,
            ))


def _infer_callable_type(fn: Closure | BuiltinFn, env: Env) -> TFn | None:
    """Infer the type signature of a callable by probing it.

    Tries calling with sample values to determine input/output types.
    Returns None if type can't be determined.
    """
    test_inputs = [
        (0.0, TNum()),
        ("", TStr()),
        (True, TBool()),
    ]

    if isinstance(fn, Closure):
        arity = len(fn.params)
    elif isinstance(fn, BuiltinFn):
        # Try to determine arity by calling with increasing args
        arity = _probe_arity(fn)
        if arity is None:
            return None
    else:
        return None

    if arity == 0:
        # Nullary function — try calling with no args
        try:
            result = apply_fn(fn, [])
            ret_type = _value_to_type(result)
            return TFn((), ret_type)
        except Exception:
            return None

    # Try each input type for the first parameter
    for test_val, test_type in test_inputs:
        try:
            args = [test_val] * arity
            result = apply_fn(fn, args)
            ret_type = _value_to_type(result)
            param_types = tuple([test_type] * arity)
            return TFn(param_types, ret_type)
        except Exception:
            continue

    return None


def _probe_arity(fn: BuiltinFn) -> int | None:
    """Determine the arity of a builtin by trying different arg counts."""
    for arity in range(4):
        try:
            args = [0.0] * arity
            apply_fn(fn, args)
            return arity
        except TypeError:
            continue
        except Exception:
            return arity  # right arity but wrong type
    return None


def _value_to_type(val: Any) -> Type:
    """Infer the SELPH type of a Python value."""
    if isinstance(val, bool):
        return TBool()
    if isinstance(val, (int, float)):
        return TNum()
    if isinstance(val, str):
        return TStr()
    if isinstance(val, list):
        if val and isinstance(val[0], str):
            return TList(TStr())
        return TList(TNum())
    if isinstance(val, Namespace):
        return TVar(0)
    return TVar(0)


# ── Multi-tree synthesis ─────────────────────────────────────────────

def multi_synthesize(
    spec: Spec,
    trees: list[Namespace],
    tree_names: list[str] | None = None,
    max_depth: int = 2,
    max_candidates: int = 50000,
    env: Env | None = None,
    enable_if: bool = False,
    include_builtins: bool = True,
) -> SynthesisResult:
    """Synthesize a program searching across multiple namespace trees.

    Each tree contributes its callable entries as synthesis components.
    Trees can be arbitrary — ops libraries, data stores, cached results,
    domain-specific knowledge, etc.

    Args:
        spec: The spec to satisfy.
        trees: List of Namespace values to search across.
        tree_names: Optional names for each tree (for traceability).
        max_depth: Maximum AST depth.
        max_candidates: Search budget.
        env: Eval environment (trees' callables will be registered here).
        enable_if: Whether to generate if-expressions.
        include_builtins: Whether to include default builtin components.
    """
    if env is None:
        env = standard_env()

    if tree_names is None:
        tree_names = [f"tree{i}" for i in range(len(trees))]

    # Extract components from all trees
    all_tree_components: list[Component] = []
    for ns, name in zip(trees, tree_names):
        comps = extract_components(ns, tree_name=name, env=env)
        all_tree_components.extend(comps)

    # Run synthesis with tree components as extra_components
    return synthesize(
        spec,
        max_depth=max_depth,
        max_candidates=max_candidates,
        extra_components=all_tree_components if all_tree_components else None,
        env=env,
        enable_if=enable_if,
    )


# ── Utilities ────────────────────────────────────────────────────────

def namespace_to_components(ns: Namespace, name: str = "",
                            env: Env | None = None) -> list[Component]:
    """Convenience: extract components from a single namespace."""
    if env is None:
        env = standard_env()
    return extract_components(ns, tree_name=name, env=env)


def trace_solution(result: SynthesisResult, trees: list[Namespace],
                   tree_names: list[str]) -> dict[str, list[str]]:
    """Trace which trees a synthesized solution draws from.

    Returns {tree_name: [paths_used]} showing which entries from
    each tree appear in the solution.
    """
    if not result.found or result.program is None:
        return {}

    usage = {name: [] for name in tree_names}

    def _scan(node: Node):
        if isinstance(node, Symbol):
            for name in tree_names:
                if node.name.startswith(name + "."):
                    path = node.name[len(name) + 1:]
                    usage[name].append(path)
        if isinstance(node, List):
            for e in node.elements:
                _scan(e)

    _scan(result.program)

    # Remove empty trees
    return {k: v for k, v in usage.items() if v}

File: selph_fast/src/multitree.rs
//! Multi-tree synthesis for SELPH.
//!
//! Allows the synthesizer to search across multiple arbitrary namespace trees
//! simultaneously. Each tree contributes callable components to the search
//! space, with paths preserved for traceability.
//!
//! A program synthesized from multiple trees might use components like:
//!   `math.double`, `str.clean`, `data.lookup`
//!
//! For efficiency, namespace lookups are resolved at synthesis time and
//! components are registered directly in the eval environment.

use std::collections::HashMap;
use crate::types::*;
use crate::eval;
use crate::synth::{
    SynthComponent, SynthResult, synthesize_with_validation,
    default_synth_components_with_trees,
    value_type_tag, TYPE_NUM, TYPE_STR, TYPE_BOOL,
};

// ── Extract components from a Namespace ──────────────────────────────

/// Recursively walk a `Value::Namespace`, extracting callable entries as
/// `SynthComponent`s with qualified names.
///
/// - `ns`: a `Value::Namespace` to walk
/// - `tree_name`: the tree's identifier (e.g. "math"), used as the prefix
/// - `prefix`: dot-separated path within the namespace (used for recursion)
/// - `env`: eval environment; callables are registered here so the
///   synthesizer can call them by name
/// - `max_depth`: maximum nesting depth to recurse into
///
/// Returns a vector of `SynthComponent`s, one per callable or constant leaf.
pub fn extract_components_from_namespace(
    ns: &Value,
    tree_name: &str,
    prefix: &str,
    env: &mut Env,
    max_depth: usize,
) -> Vec<SynthComponent> {
    let mut components = Vec::new();
    extract_recursive(ns, tree_name, prefix, env, &mut components, 0, max_depth);
    components
}

fn extract_recursive(
    ns: &Value,
    tree_name: &str,
    prefix: &str,
    env: &mut Env,
    components: &mut Vec<SynthComponent>,
    depth: usize,
    max_depth: usize,
) {
    if depth > max_depth {
        return;
    }

    let map = match ns {
        Value::Namespace(m) => m,
        _ => return,
    };

    // Sort keys for deterministic ordering
    let mut keys: Vec<&String> = map.keys().collect();
    keys.sort();

    for key in keys {
        let val = &map[key];
        let path_str = if prefix.is_empty() {
            key.clone()
        } else {
            format!("{}.{}", prefix, key)
        };
        let full_name = if tree_name.is_empty() {
            path_str.clone()
        } else {
            format!("{}.{}", tree_name, path_str)
        };

        match val {
            Value::Namespace(_) => {
                extract_recursive(
                    val, tree_name, &path_str, env, components,
                    depth + 1, max_depth,
                );
            }
            Value::Closure(..) | Value::Builtin(_) => {
                if let Some((param_types, ret_type)) = infer_callable_type(val, env) {
                    // Register the callable in the env so the synthesizer can
                    // invoke it by its qualified name.
                    env_define(env, full_name.clone(), val.clone());

                    components.push(SynthComponent {
                        name: full_name.clone(),
                        builtin: Some(full_name),
                        arity: param_types.len(),
                        ret_type,
                        param_types,
                        priority: 0.0,
                    });
                }
            }
            Value::Num(n) => {
                // Constant number — generate a component whose node is a
                // numeric literal.  No builtin needed; the synthesizer
                // will emit a Num node via the name (parseable as f64).
                let name_str = if *n == (*n as i64) as f64 {
                    format!("{}", *n as i64)
                } else {
                    format!("{}", n)
                };
                components.push(SynthComponent {
                    name: name_str,
                    builtin: None,
                    arity: 0,
                    ret_type: TYPE_NUM,
                    param_types: vec![],
                    priority: 0.0,
                });
            }
            Value::Str(s) => {
                components.push(SynthComponent {
                    name: s.clone(),
                    builtin: None,
                    arity: 0,
                    ret_type: TYPE_STR,
                    param_types: vec![],
                    priority: 0.0,
                });
            }
            Value::Bool(b) => {
                components.push(SynthComponent {
                    name: if *b { "true".to_string() } else { "false".to_string() },
                    builtin: None,
                    arity: 0,
                    ret_type: TYPE_BOOL,
                    param_types: vec![],
                    priority: 0.0,
                });
            }
            _ => {
                // Skip non-extractable values (Nil, List, RustMacro, etc.)
            }
        }
    }
}

// ── Type inference for callables ─────────────────────────────────────

/// Probe a callable value with test inputs to infer its parameter types
/// and return type as u8 type tags.
///
/// Returns `Some((param_types, ret_type))` on success, `None` if the
/// signature cannot be determined.
///
/// Strategy:
/// - For closures, arity is known from the parameter list.
/// - For builtins, arity is probed by trying 0..4 arguments.
/// - Each arity is tested with sample values (0.0, "", true) to find
///   a combination that succeeds, then the return type is inferred.
pub fn infer_callable_type(
    val: &Value,
    env: &mut Env,
) -> Option<(Vec<u8>, u8)> {
    let arity = match val {
        Value::Closure(params, _, _, _, _) => params.len(),
        Value::Builtin(_) => probe_arity(val, env)?,
        _ => return None,
    };

    if arity == 0 {
        return None; // Nullary builtins are rare, skip
    }

    // Test inputs: (value, type_tag)
    let test_inputs: [(Value, u8); 3] = [
        (Value::Num(0.0), TYPE_NUM),
        (Value::Str(String::new()), TYPE_STR),
        (Value::Bool(true), TYPE_BOOL),
    ];

    let empty_nodes: Vec<Node> = Vec::new();

    for (test_val, test_tag) in &test_inputs {
        let args: Vec<Value> = vec![test_val.clone(); arity];
        let val_clone = val.clone();
        let nodes_clone = empty_nodes.clone();
        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            let mut trial_env = eval::make_default_env();
            eval::apply(&val_clone, &args, &nodes_clone, &mut trial_env)
        }));
        if let Ok(Ok(result_val)) = result {
            let ret_type = value_type_tag(&result_val);
            let param_types = vec![*test_tag; arity];
            return Some((param_types, ret_type));
        }
    }

    None
}

/// Determine the arity of a builtin by trying increasing argument counts.
/// Uses catch_unwind because some builtins index args without bounds checks.
fn probe_arity(val: &Value, _env: &mut Env) -> Option<usize> {
    let empty_nodes: Vec<Node> = Vec::new();
    for arity in 1..=4 {
        let args: Vec<Value> = vec![Value::Num(0.0); arity];
        let val_clone = val.clone();
        let nodes_clone = empty_nodes.clone();
        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            let mut trial_env = eval::make_default_env();
            eval::apply(&val_clone, &args, &nodes_clone, &mut trial_env)
        }));
        match result {
            Ok(Ok(_)) => return Some(arity),
            Ok(Err(e)) => {
                // Type mismatch means arity is correct, just wrong test type
                if e.contains("expected") && !e.contains("not callable") {
                    return Some(arity);
                }
                continue;
            }
            Err(_) => continue, // Panic (index out of bounds) — wrong arity
        }
    }
    None
}

// ── Multi-tree synthesis ─────────────────────────────────────────────

/// Synthesize a program searching across multiple namespace trees.
///
/// Each tree contributes its callable entries as synthesis components.
/// Trees can be arbitrary — ops libraries, data stores, cached results,
/// domain-specific knowledge, etc.
///
/// Tree-extracted closures and builtins are injected as `extra_bindings`
/// into the core synthesizer's eval environment, so they can be called
/// by their qualified names during candidate evaluation.
///
/// # Arguments
/// - `inputs`: example input values
/// - `expected`: expected output values (one per input)
/// - `trees`: slice of `(name, namespace_value)` pairs
/// - `macros`: library macros (name, params, nodes, root)
/// - `max_depth`: maximum AST depth to explore
/// - `max_candidates`: search budget
/// - `include_builtins`: whether to include default builtin components
///
/// Returns a `SynthResult` with the found program (if any).
pub fn multi_synthesize(
    inputs: &[Value],
    expected: &[Value],
    trees: &[(String, Value)],
    macros: &[(String, Vec<String>, Vec<Node>, usize)],
    max_depth: usize,
    max_candidates: usize,
    include_builtins: bool,
) -> SynthResult {
    let (all_components, extra_bindings) = if include_builtins {
        default_synth_components_with_trees(macros, trees)
    } else {
        // Minimal: at least provide the input variable `x`, plus tree components
        let mut env = eval::make_default_env();
        let mut comps = vec![SynthComponent {
            name: "x".into(),
            builtin: None,
            arity: 0,
            ret_type: TYPE_NUM,
            param_types: vec![],
            priority: 100.0,
        }];
        for (name, ns_val) in trees {
            let tree_comps = extract_components_from_namespace(
                ns_val, name, "", &mut env, 5,
            );
            comps.extend(tree_comps);
        }
        // Collect extra bindings from the env
        let default_env = eval::make_default_env();
        let default_scope = default_env.last().unwrap();
        let mut bindings = Vec::new();
        if let Some(scope) = env.last() {
            for (k, v) in scope {
                if !default_scope.contains_key(k) {
                    bindings.push((k.clone(), v.clone()));
                }
            }
        }
        (comps, bindings)
    };

    synthesize_with_validation(
        &all_components,
        inputs,
        expected,
        macros,
        max_depth,
        max_candidates,
        true, // enable_if
        None, // no validation examples
        &extra_bindings,
    )
}

// ── Solution tracing ─────────────────────────────────────────────────

/// Given a synthesized solution, trace which tree names appear in it.
///
/// Scans all nodes in the solution for `Symbol` references whose names
/// start with one of the `tree_names` followed by a dot.  Returns a
/// sorted, deduplicated list of `"tree_name.path"` strings found.
///
/// # Arguments
/// - `nodes`: the node arena from the synthesis result
/// - `root`: the root index of the solution
/// - `tree_names`: the names of the trees to check for
pub fn trace_solution(
    nodes: &[Node],
    root: usize,
    tree_names: &[String],
) -> Vec<String> {
    let mut usage: HashMap<String, Vec<String>> = HashMap::new();
    for name in tree_names {
        usage.insert(name.clone(), Vec::new());
    }

    scan_node(nodes, root, tree_names, &mut usage);

    // Flatten to a list of "tree_name: path1, path2, ..." entries,
    // but only for trees that actually contributed.
    let mut result = Vec::new();
    let mut sorted_names: Vec<&String> = usage.keys().collect();
    sorted_names.sort();

    for name in sorted_names {
        let paths = &usage[name];
        if !paths.is_empty() {
            let mut sorted_paths = paths.clone();
            sorted_paths.sort();
            sorted_paths.dedup();
            for path in sorted_paths {
                result.push(format!("{}.{}", name, path));
            }
        }
    }

    result
}

/// Recursively scan a node tree for symbol references from named trees.
fn scan_node(
    nodes: &[Node],
    idx: usize,
    tree_names: &[String],
    usage: &mut HashMap<String, Vec<String>>,
) {
    if idx >= nodes.len() {
        return;
    }
    match &nodes[idx] {
        Node::Symbol(name) => {
            for tree_name in tree_names {
                let prefix = format!("{}.", tree_name);
                if name.starts_with(&prefix) {
                    let path = name[prefix.len()..].to_string();
                    if let Some(paths) = usage.get_mut(tree_name) {
                        paths.push(path);
                    }
                }
            }
        }
        Node::App(children) => {
            for &child in children {
                scan_node(nodes, child, tree_names, usage);
            }
        }
        Node::If(c, t, e) => {
            scan_node(nodes, *c, tree_names, usage);
            scan_node(nodes, *t, tree_names, usage);
            scan_node(nodes, *e, tree_names, usage);
        }
        Node::Lambda(_, body) => {
            scan_node(nodes, *body, tree_names, usage);
        }
        Node::Let(bindings, body) => {
            for (_, val_idx) in bindings {
                scan_node(nodes, *val_idx, tree_names, usage);
            }
            scan_node(nodes, *body, tree_names, usage);
        }
        _ => {}
    }
}

// ── Convenience ──────────────────────────────────────────────────────

/// Extract components from a single namespace (convenience wrapper).
pub fn namespace_to_components(
    ns: &Value,
    name: &str,
) -> Vec<SynthComponent> {
    let mut env = eval::make_default_env();
    extract_components_from_namespace(ns, name, "", &mut env, 5)
}

// ── Tests ────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    fn empty_ns() -> Value {
        Value::Namespace(HashMap::new())
    }

    fn make_ns(entries: Vec<(&str, Value)>) -> Value {
        let mut map = HashMap::new();
        for (k, v) in entries {
            map.insert(k.to_string(), v);
        }
        Value::Namespace(map)
    }

    // ── extract_components_from_namespace ─────────────────────────────

    #[test]
    fn test_extract_empty_namespace() {
        let ns = empty_ns();
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(&ns, "test", "", &mut env, 5);
        assert!(comps.is_empty());
    }

    #[test]
    fn test_extract_numeric_constants() {
        let ns = make_ns(vec![
            ("pi", Value::Num(3.14)),
            ("zero", Value::Num(0.0)),
        ]);
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(&ns, "math", "", &mut env, 5);
        assert_eq!(comps.len(), 2);
        for comp in &comps {
            assert_eq!(comp.arity, 0);
            assert_eq!(comp.ret_type, TYPE_NUM);
        }
    }

    #[test]
    fn test_extract_string_constants() {
        let ns = make_ns(vec![
            ("greeting", Value::Str("hello".to_string())),
        ]);
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(&ns, "data", "", &mut env, 5);
        assert_eq!(comps.len(), 1);
        assert_eq!(comps[0].ret_type, TYPE_STR);
        assert_eq!(comps[0].name, "hello");
    }

    #[test]
    fn test_extract_bool_constants() {
        let ns = make_ns(vec![
            ("flag", Value::Bool(true)),
        ]);
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(&ns, "config", "", &mut env, 5);
        assert_eq!(comps.len(), 1);
        assert_eq!(comps[0].ret_type, TYPE_BOOL);
        assert_eq!(comps[0].name, "true");
    }

    #[test]
    fn test_extract_builtins() {
        let ns = make_ns(vec![
            ("double", Value::Builtin("add".to_string())),
        ]);
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(&ns, "ops", "", &mut env, 5);
        // "add" is a known builtin taking 2 nums, so it should be extracted
        assert_eq!(comps.len(), 1);
        assert_eq!(comps[0].name, "ops.double");
        assert_eq!(comps[0].arity, 2);
        assert_eq!(comps[0].ret_type, TYPE_NUM);
    }

    #[test]
    fn test_extract_nested_namespace() {
        let inner = make_ns(vec![
            ("value", Value::Num(42.0)),
        ]);
        let outer = make_ns(vec![
            ("deep", inner),
        ]);
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(&outer, "root", "", &mut env, 5);
        assert_eq!(comps.len(), 1);
        assert_eq!(comps[0].ret_type, TYPE_NUM);
    }

    #[test]
    fn test_extract_max_depth_limit() {
        let deep = make_ns(vec![("val", Value::Num(1.0))]);
        let mid = make_ns(vec![("inner", deep)]);
        let top = make_ns(vec![("mid", mid)]);
        let mut env = eval::make_default_env();
        // Depth 0 means only the top-level namespace entries (which is
        // itself a namespace, so nothing extracted)
        let comps = extract_components_from_namespace(&top, "t", "", &mut env, 0);
        // At depth 0, we enter the top namespace (depth=0) and see "mid"
        // which is a namespace. Recursing would be depth=1 > max_depth=0,
        // so it stops.
        assert!(comps.is_empty());
    }

    #[test]
    fn test_extract_non_namespace_is_noop() {
        let mut env = eval::make_default_env();
        let comps = extract_components_from_namespace(
            &Value::Num(42.0), "t", "", &mut env, 5,
        );
        assert!(comps.is_empty());
    }

    // ── infer_callable_type ──────────────────────────────────────────

    #[test]
    fn test_infer_builtin_add() {
        let val = Value::Builtin("add".to_string());
        let mut env = eval::make_default_env();
        let result = infer_callable_type(&val, &mut env);
        assert!(result.is_some());
        let (params, ret) = result.unwrap();
        assert_eq!(params, vec![TYPE_NUM, TYPE_NUM]);
        assert_eq!(ret, TYPE_NUM);
    }

    #[test]
    fn test_infer_builtin_string_upper() {
        let val = Value::Builtin("string-upper".to_string());
        let mut env = eval::make_default_env();
        let result = infer_callable_type(&val, &mut env);
        assert!(result.is_some());
        let (params, ret) = result.unwrap();
        assert_eq!(params, vec![TYPE_STR]);
        assert_eq!(ret, TYPE_STR);
    }

    #[test]
    fn test_infer_builtin_not() {
        let val = Value::Builtin("not".to_string());
        let mut env = eval::make_default_env();
        let result = infer_callable_type(&val, &mut env);
        assert!(result.is_some());
        let (params, ret) = result.unwrap();
        assert_eq!(params, vec![TYPE_BOOL]);
        assert_eq!(ret, TYPE_BOOL);
    }

    #[test]
    fn test_infer_non_callable() {
        let val = Value::Num(42.0);
        let mut env = eval::make_default_env();
        assert!(infer_callable_type(&val, &mut env).is_none());
    }

    #[test]
    fn test_infer_closure() {
        // Create a simple closure: (lambda (a b) (add a b))
        let nodes = vec![
            Node::Symbol("add".to_string()),   // 0
            Node::Symbol("a".to_string()),      // 1
            Node::Symbol("b".to_string()),      // 2
            Node::App(vec![0, 1, 2]),           // 3: (add a b)
        ];
        let env = eval::make_default_env();
        let closure = Value::Closure(
            vec!["a".to_string(), "b".to_string()],
            3,          // body index
            env.clone(),
            nodes,
            None,
        );
        let mut test_env = eval::make_default_env();
        let result = infer_callable_type(&closure, &mut test_env);
        assert!(result.is_some());
        let (params, ret) = result.unwrap();
        assert_eq!(params.len(), 2);
        assert_eq!(params[0], TYPE_NUM);
        assert_eq!(ret, TYPE_NUM);
    }

    // ── trace_solution ───────────────────────────────────────────────

    #[test]
    fn test_trace_empty_solution() {
        let nodes = vec![Node::Num(42.0)];
        let tree_names = vec!["math".to_string(), "str".to_string()];
        let result = trace_solution(&nodes, 0, &tree_names);
        assert!(result.is_empty());
    }

    #[test]
    fn test_trace_with_tree_references() {
        let nodes = vec![
            Node::Symbol("math.double".to_string()),  // 0
            Node::Symbol("x".to_string()),             // 1
            Node::App(vec![0, 1]),                     // 2: (math.double x)
        ];
        let tree_names = vec!["math".to_string(), "str".to_string()];
        let result = trace_solution(&nodes, 2, &tree_names);
        assert_eq!(result, vec!["math.double".to_string()]);
    }

    #[test]
    fn test_trace_multiple_trees() {
        let nodes = vec![
            Node::Symbol("str.clean".to_string()),     // 0
            Node::Symbol("x".to_string()),              // 1
            Node::App(vec![0, 1]),                      // 2: (str.clean x)
            Node::Symbol("math.double".to_string()),    // 3
            Node::App(vec![3, 2]),                      // 4: (math.double (str.clean x))
        ];
        let tree_names = vec!["math".to_string(), "str".to_string()];
        let result = trace_solution(&nodes, 4, &tree_names);
        assert_eq!(result, vec![
            "math.double".to_string(),
            "str.clean".to_string(),
        ]);
    }

    #[test]
    fn test_trace_nested_paths() {
        let nodes = vec![
            Node::Symbol("lib.math.trig.sin".to_string()),  // 0
            Node::Symbol("x".to_string()),                    // 1
            Node::App(vec![0, 1]),                            // 2
        ];
        let tree_names = vec!["lib".to_string()];
        let result = trace_solution(&nodes, 2, &tree_names);
        assert_eq!(result, vec!["lib.math.trig.sin".to_string()]);
    }

    #[test]
    fn test_trace_if_expression() {
        let nodes = vec![
            Node::Symbol("ops.check".to_string()),  // 0
            Node::Symbol("x".to_string()),           // 1
            Node::App(vec![0, 1]),                   // 2: (ops.check x)
            Node::Symbol("ops.then".to_string()),    // 3
            Node::App(vec![3, 1]),                   // 4: (ops.then x)
            Node::Num(0.0),                          // 5
            Node::If(2, 4, 5),                       // 6: (if (ops.check x) (ops.then x) 0)
        ];
        let tree_names = vec!["ops".to_string()];
        let result = trace_solution(&nodes, 6, &tree_names);
        assert_eq!(result, vec![
            "ops.check".to_string(),
            "ops.then".to_string(),
        ]);
    }

    #[test]
    fn test_trace_deduplicates() {
        let nodes = vec![
            Node::Symbol("math.add".to_string()),   // 0
            Node::Symbol("x".to_string()),            // 1
            Node::App(vec![0, 1, 1]),                 // 2: (math.add x x)
            // Duplicate symbol reference
            Node::Symbol("math.add".to_string()),    // 3
            Node::App(vec![3, 2, 2]),                 // 4: (math.add (math.add x x) (math.add x x))
        ];
        let tree_names = vec!["math".to_string()];
        let result = trace_solution(&nodes, 4, &tree_names);
        // "math.add" should appear only once despite multiple occurrences
        assert_eq!(result, vec!["math.add".to_string()]);
    }

    // ── multi_synthesize ─────────────────────────────────────────────

    #[test]
    fn test_multi_synthesize_with_builtin_tree() {
        // Build a tree that contains "add" as a component named "ops.plus"
        let ns = make_ns(vec![
            ("plus", Value::Builtin("add".to_string())),
        ]);
        let trees = vec![("ops".to_string(), ns)];
        let macros: Vec<(String, Vec<String>, Vec<Node>, usize)> = Vec::new();

        // Simple spec: double(x) = x + x
        // This should be findable with the default builtins alone
        let inputs = vec![Value::Num(1.0), Value::Num(3.0), Value::Num(5.0)];
        let expected = vec![Value::Num(2.0), Value::Num(6.0), Value::Num(10.0)];

        let result = multi_synthesize(
            &inputs, &expected, &trees, &macros,
            2, 50000, true,
        );

        assert!(result.found);
    }

    #[test]
    fn test_multi_synthesize_no_trees() {
        let trees: Vec<(String, Value)> = vec![];
        let macros: Vec<(String, Vec<String>, Vec<Node>, usize)> = Vec::new();

        // Simple identity: f(x) = x
        let inputs = vec![Value::Num(1.0), Value::Num(2.0)];
        let expected = vec![Value::Num(1.0), Value::Num(2.0)];

        let result = multi_synthesize(
            &inputs, &expected, &trees, &macros,
            1, 10000, true,
        );

        assert!(result.found);
    }

    // ── namespace_to_components ───────────────────────────────────────

    #[test]
    fn test_namespace_to_components_convenience() {
        let ns = make_ns(vec![
            ("val", Value::Num(99.0)),
            ("flag", Value::Bool(false)),
        ]);
        let comps = namespace_to_components(&ns, "cfg");
        assert_eq!(comps.len(), 2);
    }
}
